### PR TITLE
Feature/gh 12143

### LINKED ISSUE #12143: Implement functionalities for multi-d variants selectors on PDP

### PR DESCRIPTION
**Pull request by dydome**
close #12143

### CODE REVIEW

File: feature-libs/product/multi-d/components/variants-multi-dimensional-selector/variants-multi-dimensional-selector.component.html [hunk 1]
```diff
@@ -1,23 +1,31 @@
 <ng-container *ngIf="product">
   <div class="variant-generic-selector">
-    <ng-container *ngFor="let variant of variants">
-      <span>{{ variant[0].parentVariantCategory.name }}</span>
+    <ng-container *ngFor="let variant of multiDimensionalService.getVariants()">
+      <label>{{ variant[0].parentVariantCategory.name }}</label>
 
       <div
-        *ngIf="variantHasImages(variant); else noImages"
+        *ngIf="multiDimensionalService.variantHasImages(variant); else noImages"
```
Comment (psmul) [inline feature-libs/product/multi-d/components/variants-multi-dimensional-selector/variants-multi-dimensional-selector.component.html]: let's avoid using serviced strictly in templates. Our logic controller (ts component) should be serving as middleware here.

File: feature-libs/product/multi-d/components/variants-multi-dimensional-selector/variants-multi-dimensional-selector.component.ts [hunk 2]
```diff
@@ -46,50 +51,40 @@ export class VariantsMultiDimensionalSelectorComponent implements OnInit {
...
-    this.variants = [];
-
-    const levels = Array.from(
-      { length: this.product?.categories?.length },
-      (_, k) => k + 1
-    );
-
-    let productMatrix = JSON.parse(JSON.stringify(this.product.variantMatrix));
+  getVariantOptionImages(variantOptionQualifier: VariantOptionQualifier[]) {
+    const elements = JSON.parse(JSON.stringify(variantOptionQualifier));
```
Comment (psmul) [inline feature-libs/product/multi-d/components/variants-multi-dimensional-selector/variants-multi-dimensional-selector.component.ts]: what's the puprose of parse and stringify here?

File: feature-libs/product/multi-d/components/variants-multi-dimensional-selector/variants-multi-dimensional-selector.component.html [hunk 3]
```diff
@@ -1,23 +1,31 @@
...
       >
+        <!-- TO DO check 3 element for image -->
         <cx-media
           *ngFor="let cat of variant"
           (click)="changeVariant(cat.variantOption.code)"
-          [container]="cat.variantOption?.variantOptionQualifiers[0]?.image"
+          [container]="
+            getVariantOptionImages(cat.variantOption?.variantOptionQualifiers)
+          "
+          [title]="cat.variantValueCategory.name"
```
Comment (psmul) [inline feature-libs/product/multi-d/components/variants-multi-dimensional-selector/variants-multi-dimensional-selector.component.html]: can we change cat to category? Just for readability and clear purpose

File: feature-libs/product/multi-d/components/variants-multi-dimensional-selector/variants-multi-dimensional-selector.component.ts [hunk 4]
```diff
@@ -46,50 +51,40 @@ export class VariantsMultiDimensionalSelectorComponent implements OnInit {
...
-        productMatrix =
-          productMatrix[currentLevelProductVariantIndex]?.elements;
-      }
+    elements.forEach((element: any) => {
+      const imageObject = {
+        [element.image?.format as any]: Object.assign(defaultImageObject, {
+          format: element.image?.format,
+          url: this.getBaseUrl() + element.image?.url,
+        }),
+      } as Image;
```
Comment (psmul) [inline feature-libs/product/multi-d/components/variants-multi-dimensional-selector/variants-multi-dimensional-selector.component.ts]: Why do we need to do type casting here? Can we explore alternatives? Maybe we can use Set?

File: feature-libs/product/multi-d/components/variants-multi-dimensional/variants-multi-dimensional.component.html [hunk 5]
```diff
@@ -1,4 +1,7 @@
-<div *ngIf="product$ | async as product" class="variants-section">
+<div
+  *ngIf="multiDimensionalService.product$ | async as product"
```
Comment (psmul) [inline feature-libs/product/multi-d/components/variants-multi-dimensional/variants-multi-dimensional.component.html]: Let's avoid using services strictly in templates

File: feature-libs/product/multi-d/occ/config/default-occ-variants-multidimensional-config.ts [hunk 6]
```diff
@@ -4,8 +4,7 @@ export const defaultOccProductVariantsMultidimensionalConfig: OccConfig = {
   backend: {
     occ: {
       endpoints: {
-        product:
-          'products/${productCode}?fields=name,averageRating,stock(DEFAULT),description,summary,availableForPickup,code,url,price(DEFAULT),numberOfReviews,manufacturer,categories(FULL),priceRange,multidimensional,tags,images(FULL),purchasable,baseOptions(DEFAULT),baseProduct,variantOptions(DEFAULT),variantMatrix(DEFAULT)',
+        product: 'products/${productCode}?fields=FULL',
```
Comment (psmul) [inline feature-libs/product/multi-d/occ/config/default-occ-variants-multidimensional-config.ts:7]: Is this intended?

**Review by psmul — CHANGES_REQUESTED** [review]

File: feature-libs/product/multi-d/occ/config/default-occ-variants-multidimensional-config.ts (hunk 6 above)
Comment (dydome) [reply feature-libs/product/multi-d/occ/config/default-occ-variants-multidimensional-config.ts:7]: yes we need this to full until API will be stable

File: feature-libs/product/multi-d/components/variants-multi-dimensional-selector/variants-multi-dimensional-selector.component.ts [hunk 7]
```diff
@@ -50,30 +57,25 @@ export class VariantsMultiDimensionalSelectorComponent implements OnInit {
             cxRoute: 'product',
             params: product,
           });
-          this.product = product;
+          this.product = product as Product;
```
Comment (psmul) [inline feature-libs/product/multi-d/components/variants-multi-dimensional-selector/variants-multi-dimensional-selector.component.ts:60]: this seems unnecessary as we provide type along with @Input()

**Review by psmul — CHANGES_REQUESTED** [review]

**Review by psmul — APPROVED** [review]

**Pull request merged**